src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala

```diff
@@ -188,25 +194,47 @@ class TaskTracker @Inject() (
     log.debug(s"Fetching app from store $appId")
     val names = timedRead { state.names().get.asScala.toSet }
     val tasks = TrieMap[String, MarathonTask]()
-    val taskKeys = names.filter(name => name.startsWith(PREFIX + appId.safePath + ID_DELIMITER))
+    val taskKeys = names.filter(name => name.startsWith(getKeyPrefix(appId)))
     for {
       taskKey <- taskKeys
-      task <- fetchTask(taskKey)
+      task <- fetchTaskByKey(appId, taskKey)
     } tasks += (task.getId -> task)
 
     new InternalApp(appId, tasks, false)
   }
 
-  def fetchTask(appId: PathId, taskId: String): Option[MarathonTask] =
-    fetchTask(getKey(appId, taskId))
+  def fetchTaskById(appId: PathId, taskId: String): Option[MarathonTask] =
+    fetchTaskByKey(appId, getKey(appId, taskId))
 
-  private[tasks] def fetchTask(taskKey: String): Option[MarathonTask] = {
-    val bytes = fetchFromState(taskKey).value
-    if (bytes.length > 0) {
-      val source = new ObjectInputStream(new ByteArrayInputStream(bytes))
-      deserialize(taskKey, source)
+  private[tasks] def fetchTaskByKey(appId: PathId, taskKey: String): Option[MarathonTask] = {
+    // Search the in-memory storage for the supplied task.
+    def inMemory(): Option[MarathonTask] = {
+      val taskId = taskKey.stripPrefix(getKeyPrefix(appId))
+      apps.get(appId).flatMap { app =>
+        app.tasks.get(taskId)
+      }
     }
-    else None
+
+    // Search the non-local storage for the supplied task.
+    def fromPersistentStorage(): Option[MarathonTask] = {
+      val bytes = fetchFromState(taskKey).value
+      if (bytes.length > 0) {
+        val source = new ObjectInputStream(new ByteArrayInputStream(bytes))
+        val taskOption = deserialize(taskKey, source)
+
+        for (task <- taskOption) {
+          // Update the in-memory storage.
+          val internalApp = apps.getOrElseUpdate(appId, new InternalApp(appId, TrieMap(), false))
+          internalApp.tasks += (taskKey -> task)
```

Comment:
This should be putIfAbsent, in case another thread added this task in the meantime.